pickhardtpayments/Payment.py

```diff
@@ -175,9 +261,449 @@ def successful(self) -> bool:
+        logger.debug("planned fee: \t\t{:8.0f} sat".format(self.total_fees / 1000))
+        logger.debug("fees for in_flights:\t{:8.0f} sat".format(self.filter_fees(AttemptStatus.INFLIGHT) / 1000))
+        if self._end_time:
+            logger.debug("Runtime of flow computation: {:4.2f} sec".format(self._end_time - self._start_time))
+
+    def execute(self) -> int:
+        """
+        Executes the Payment.
+        This is the last step in the payment loop. The current in_flight attempts are settled. To achieve this,
+        settle_attempt() is called on the Oracle Network as well as on the Uncertainty Network to settle the attempt.
+        The in_flights are removed in OracleNetworks and UncertaintyNetwork for this Attempt and the channel balances
+        - or the belief about the channel balances - are adjusted.
+
+        """
+        logger.info("Executing Payment...")
+        logger.debug("settling {} Attempts:".format(len(list(self.filter_attempts(AttemptStatus.INFLIGHT)))))
+        for attempt in self.filter_attempts(AttemptStatus.INFLIGHT):
+            for channel in attempt.path:
+                logger.debug("(04) {}-{} Uncertainty Range:\t\t[{:>10,} ; {:>10,}]\t\tinflight {:>10,};"
+                             "\tcond_cap: {:>10,}".format(channel.src[:4],
+                                                          channel.dest[:4],
+                                                          channel.min_liquidity,
+                                                          channel.max_liquidity,
+                                                          channel.in_flight,
+                                                          channel.conditional_capacity))
+                return_channel = self.uncertainty_network.get_channel(channel.dest, channel.src,
+                                                                      channel.short_channel_id)
+                if return_channel:
+                    logger.debug("(05) {}-{} Uncertainty Range:\t\t[{:>10,} ; {:>10,}]\t\tinflight {:>10,};"
+                                 "\tcond_cap: {:>10,}".format(return_channel.src[:4],
+                                                              return_channel.dest[:4],
+                                                              return_channel.min_liquidity,
+                                                              return_channel.max_liquidity,
+                                                              return_channel.in_flight,
+                                                              return_channel.conditional_capacity))
+
+            try:
+                logger.debug("settling OracleNetwork...")
+                self._oracle_network.settle_attempt(attempt)  # updates in_flights and liquidity in both paths
+                logger.debug("settled. settling UncertaintyNetwork...")
+                self._uncertainty_network.settle_attempt(attempt)  # removes along path
```

Comment:
it feels as if oracle and uncertainty network settle_attempt should happen atomically. I think it does not matter for now but on the long term this is a bit of an accuracy